rfcs/2024-05-21-ui-component-model.md

```diff
@@ -0,0 +1,140 @@
+        - [overridable blocks](https://mustache.github.io/mustache.5.html#Blocks)
+- Components have **[high locality of behavior](https://github.com/gordonbrander/generative-ui-playbook?tab=readme-ov-file#llms-work-best-with-high-locality)** (P1)
+    - All component behavior is colocated, including structure, style, and behavior
+- Components are relatively **small**
+    - LLMs will have higher accuracy generating small islands of interactivity vs whole apps
+    - Small composable components are easier to understand and debug
+- UI **templates** are **pure functions** (P1)
+    - Templates take inputs, and output a UI tree and events
+    - Templates produce a UI tree that is easy for the runtime to analyze and sanitize (probably a VDOM, probably not raw DOM).
+- Components are renderable to web (P1)
+    - Other platforms may be supported in future, but web platform is primary
+
+Soft goals:
+
+- UI templates are static (P3)
+    - They are compiled once at program start, and produce a static tree with specific “binding points” in the tree, where dynamic values and dynamic lists are rendered.
+    - This may have a performance advantage over a totally dynamic VDOM tree, since it would allow us to analyze and enforce policies on the tree once, rather than after every render
+
+### Non-goals
+
+- Separation of concerns. At odds with high locality of behavior.
+
+## Proposal
+
+## Alternatives
+
+### React-style functional components
+
+### Stateless templates
+
+Borrowing ideas from Mustache, Vue, and Svelte, we could separate logic from template. This would make the template a pure function. It would also encourage factoring out the logic into signal transformations.
+
+Key features:
+
+- All domain logic is pulled out of the template and is performed as signal graph transformations outside of the template.
+- Signals, values, and callbacks are exported from the “script” portion of the module
+- Mustache-style static templates
+    - Ordinary values are rendered statically
+    - Signals are rendered reactively
+    - Callbacks can be used to send messages up from the template
+
+A simple counter example:
+
+```html
+<script>
+  const [count, setCount] = signal(0)
+  export count
+
+  const [clicks, setClicks] = stream()
+  export setClicks
+
+  clicks.sink(_ => setCount(count() + 1))
+</script>
+
+<template>
+  <a onclick="{{setClicks}}">The count is: {{count}}</a>
+</template>
+```
+
+Under the hood, the system might be doing something like this:
+
+```js
+// ...Somewhere in the runtime, invisible to the module
+
+// System somehow compiles the component definition
+const {env, template} = compile(component)
+
+// `template` contains the string contents of the template block.
+// `env` contains the exports from the script block, e.g.
+// const {setClicks, count} = env
+
+// Template is turned into a UI tree with dynamic bindings
+// at specific locations in the tree
+const vdom = populate(template, env)
```

Comment:
interesting, so this could be seen in two ways:

1) this is internal to the module (and the runtime just understands vdom + bound event streams)

2) this is actually two modules wrapped in a mini recipe
 - the script part which is exports `count` and `setClicks`
 - the template part, which reads `count` and receives `setClicks` so it can issue events on it

it looks a bit like (2), but i have a few questions on the lifecycle then (it might actually be easier for example for the template to export `setClicks` and the code module to read from it.

Reply:
As noted in the code snippet, I'm attempting to show hypothetical "code behind the curtain" in the runtime to illustrate that this approach supports:

- compiling a static template
- importing signals and wiring them up to the template
- sanitizing the resulting UI description tree
- rendering that tree without the module having any direct access to the DOM

This code snippet was not intended to illustrate what is happening at the module or recipe level, just the kind of thing that might be happening behind the scenes in the runtime when validating/rendering the output.